### PR TITLE
* #229: Added missing parsing for the `aggregationType` field of the request

### LINKED ISSUE #229: Add implementation for `aggregationType` field parsing of the pushdown request
## Problem 

In the parsing of the pushdown query, we don't parse aggregationType field currently. `group_by` statement is parsed when present, but `single_group` value is ignored. This leads to incorrect results of some query executions.

From our documentation:

aggregationType: An optional element, set if aggregation is requested. Either group_by or single_group, if an aggregate function is used but no group by.

## How to reproduce an error:

```
create virtual schema vs using adapter.jdbc_adapter
with connection_name = 'SYS_CONNECTION' 
sql_dialect='EXASOL' schema_name = 'TEST';

-- This returns two rows instead of 1
select 1 as bar from (select sum(x) x from vs.t1); 
```
### Example of a pushdown request:

```
 "pushdownRequest" :
        {
                "aggregationType" : "single_group",
                "from" :
                {
                        "name" : "T1",
                        "type" : "table"
                },
                "selectList" :
                [
                        {
                                "type" : "literal_null"
                        }
                ],
                "type" : "select"
        },
```

### PR DESCRIPTION
**Pull request by AnastasiiaSergienko**
Closes #229

### CODE REVIEW

File: src/main/java/com/exasol/adapter/request/renderer/PushdownSqlRenderer.java [hunk 1]
```diff
@@ -65,6 +65,19 @@ private void addIfPresent(final SqlNode node, final String key, final JsonObject
             }
         }
 
+        private void addGroupByIfPresent(final SqlExpressionList groupBy, final JsonObjectBuilder builder)
+                throws AdapterException {
+            if (groupBy != null) {
+                final JsonValue groupByJson = groupBy.accept(this);
+                if (groupByJson.toString().contains("single_group")) {
```
Comment (jakobbraun) [inline src/main/java/com/exasol/adapter/request/renderer/PushdownSqlRenderer.java:72]: that looks dangerous. The json could also contain `single_group` as a column name.

File: src/main/java/com/exasol/adapter/request/renderer/PushdownSqlRenderer.java [hunk 2]
```diff
@@ -79,8 +92,14 @@ private JsonArray convertListOfNodes(final List<SqlNode> expressions) throws Ada
         }
 
         @Override
-        public JsonArray visit(final SqlGroupBy groupBy) throws AdapterException {
-            return convertListOfNodes(groupBy.getExpressions());
+        public JsonValue visit(final SqlGroupBy groupBy) throws AdapterException {
+            if (groupBy.getExpressions().size() == 1
+                    && groupBy.getExpressions().get(0).getType() == SqlNodeType.LITERAL_STRING
+                    && ((SqlLiteralString) groupBy.getExpressions().get(0)).getValue().equals("a")) {
```
Comment (jakobbraun) [inline src/main/java/com/exasol/adapter/request/renderer/PushdownSqlRenderer.java:98]: what is this condition for? Why do you check if the value is equal to `a`?

File: src/main/java/com/exasol/adapter/request/renderer/PushdownSqlRenderer.java [hunk 3]
```diff
@@ -79,8 +92,14 @@ private JsonArray convertListOfNodes(final List<SqlNode> expressions) throws Ada
...
+                return JSON.createObjectBuilder().add("aggregationType", "single_group").build();
```
Comment (jakobbraun) [inline src/main/java/com/exasol/adapter/request/renderer/PushdownSqlRenderer.java:99]: is there a reason why you don't use `AGGREGATION_TYPE` here?

File: src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java [hunk 4]
```diff
@@ -192,8 +198,30 @@ private SqlNode parseLiteralBool(final JsonObject expression) {
         return sqlNodes;
     }
 
-    private SqlGroupBy parseGroupBy(final JsonArray groupBy) {
-        return groupBy == null ? null : new SqlGroupBy(parseExpressionList(groupBy));
+    private SqlGroupBy parseGroupBy(final JsonObject select, final SqlSelectList selectList) {
+        if (hasSingleGroupAggregation(select) && !hasAggregateFunction(selectList)) {
+            // If the aggregationType is single_group and there is no an aggregate function,
+            // we limit the result to a single row.
+            return new SqlGroupBy(List.of(new SqlLiteralString("a")));
```
Comment (jakobbraun) [inline src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java:205]: I think that functionallity does not belong into the parser. As far as I unsersdtand, it's alreaddy an implementation for handling these types of groups. This structure should in ,y oppinion be a representation of the JSON format. Different virtual schemas can then handle it differently.
You could so this transformation in a sepearate class called for example `SingleGroupTransformter`

**Review by jakobbraun — CHANGES_REQUESTED** [review]

**Review by narmion — APPROVED** [review]
Perhaps you can add the two additional test cases I suggested.
Besides that, this is ready to merge after you have addressed the open comments.
Thank you for fixing that :pray:

**Review by jakobbraun — APPROVED** [review]

**Pull request merged**